Use PhantomData to ensure Context isn't dropped while in use

GabrielMajeri opened the pull request:
Until now, the code kept some unused references to the udev Context structure, to ensure it's not prematurely dropped while the Device objects are alive.

This pull request replaces that usage with the `PhantomData` marker trait, which ensures the lifetime of the Context contains the lifetimes of the child Devices.

GabrielMajeri committed:
Use PhantomData to ensure Contex is alive while in use

dcuddeback commented:
Could you explain what problem this is solving? The context reference was already ensuring that it wasn't dropped prematurely. Is there a bug? If so, please file a bug report with steps to reproduce so I can verify that this fixes it.

GabrielMajeri commented:
@dcuddeback It is reducing the memory usage for each struct. Before it was `2 * sizeof(usize)`, now it is only `sizeof(usize)`, since [`PhantomData`](https://doc.rust-lang.org/std/marker/struct.PhantomData.html) doesn't actually use up space.

It also makes the intent of the code clearer, we don't need the context reference, we just need to ensure the object isn't prematurely dropped.

This is not fixing a bug, it is just refactoring the code.

dcuddeback commented:
@GabrielMajeri Ah, that makes sense. Sorry, it wasn't clear from the original description; I thought you were describing a bug. I made a similar change to `libusb-rs`. Here's that commit for reference: https://github.com/dcuddeback/libusb-rs/commit/0eaf40f4f1a50913e157cfc429cd6af00a45f094.

I think this would be a valuable thing to merge. Would you mind removing the stylistic changes from the PR? I.e., just change the type of the field and maybe rename from `_context` to `context`. I'd rather keep the same field ordering, explicit lifetimes, and brace structure. Also, as in the libusb commit that I linked to above, I think the type `PhantomData<&'a Context>` is more clear about the field's purpose than `PhantomData<&'a ()>`.

Did you change to `PhantomData` everywhere that it could be used? I didn't see any others in a quick search through the code, but wondering if you looked for other, too, because ideally we would change to `PhantomData` everywhere at once.

GabrielMajeri committed:
Reorder some fields

GabrielMajeri committed:
Undo indentation change

GabrielMajeri committed:
Simplify new `device::new` function

GabrielMajeri committed:
Use PhantomData in Monitor struct

GabrielMajeri commented:
@dcuddeback Changes:
- undone the indentation change
- kept the field order and field names as the original ones
- modified the `Monitor` struct to use `PhantomData`
- simplified the `device::new` function, it no longer takse that unused `context` parameter.

dcuddeback commented on src/device.rs:
```diff
@@ -3,22 +3,23 @@ use std::str;
 use ::handle::*;
 
-pub fn new(context: &Context, device: *mut ::ffi::udev_device) -> Device {
+pub fn new<'a>(device: *mut ::ffi::udev_device) -> Device<'a> {
```
How would the compiler know what lifetime to assign to `'a` without taking any references with `'a`? I think `context` is still needed here, but it can be changed to `PhantomData` like this change in libusb-rs: https://github.com/dcuddeback/libusb-rs/commit/0eaf40f4f1a50913e157cfc429cd6af00a45f094#diff-8f9dfd91ea99cdce77b0fe65d6704a02R200.

GabrielMajeri replied:
Type/lifetime inference would be done at the call site.

For example, if I write a function
```rust
fn get_device() -> Device<'static> {
  ::device::new(...)
}
```
would create a new `Device` with a static lifetime.

In this crates' case, in both cases where this function is used, the lifetime can be inferred.

- In `device_from_syspath` I modified the return type to clearly describe the lifetime.
- In `monitor.rs` it can be inferred from the returned `Event`'s lifetime.

dcuddeback replied:
I see. That makes sense, but I think that's more error-prone than propagating an existing lifetime. It'd be a lot easier for me to maintain if the `PhantomData` is created once based on the context's lifetime and then simply copied to derivative objects.

dcuddeback commented on src/device.rs:
```diff
@@ -101,7 +102,7 @@ impl<'a> Device<'a> {
 
             Some(Device {
-                _context: self._context,
+                context: PhantomData,
```
Why not continue to pass `self.context` to ensure that the correct lifetime is propagated to new objects? You can refer to the same commit in libusb-rs to see how this works: https://github.com/dcuddeback/libusb-rs/commit/0eaf40f4f1a50913e157cfc429cd6af00a45f094#diff-2c70df66cd094bd49f190698078ddb18R35.

GabrielMajeri replied:
Type inference again, since there is only one reference / lifetime going in (the one of `&self`) and only one going out (the one of `Device<'a>`),  the compiler can match them up.

dcuddeback commented:
@GabrielMajeri Thanks. I like the new changes. It looks like you found more places to use `PhantomData`, too. I just want to point out that the `context` parameter in `device::new` isn't unused. It was being used for its lifetime. I made a couple inline comments regarding how lifetimes are propagated. I keep referring back to equivalent changes in libusb-rs, because that commit has been tested.

GabrielMajeri commented:
@dcuddeback With Rust's static lifetime evaluation we can be confident that if it compiles, then it works.

For example, testing the new changes:

```rust
let device;

{
  let context = libudev::Context::new().unwrap();

  let mut devices = libudev::Enumerator::new(&context).unwrap()
    .scan_devices().unwrap();

  device = devices.next();
}
```

This code fails to compile, Rust complains that the context would be dropped too early.

I think it's safe to merge this code.

dcuddeback commented:
> With Rust's static lifetime evaluation we can be confident that if it compiles, then it works.

That's not true in the presence of raw pointers, which this crate uses extensively. From the user's point of view, you're correct if you assume that the library implements its lifetimes correctly. My comments are from the point of view of maintaining this library, where I do have to think about lifetimes. At the end of the day, both approaches work, but I'd like to merge what's easier for me to maintain.

dcuddeback commented:
This is no longer needed. I'm going in a different direction for v0.3. Commit f36b5a04b7ecd732f01945672e861a7f9472b049 solves the same issue.
